src/gui/widgets/typed_debugger.h

```diff
@@ -0,0 +1,171 @@
+     * Class structs.
+     */
+
+    // Represents the type-name-size tuple for either a struct field or a function argument.
+    struct FieldOrArgumentData {
+        std::string type;
+        std::string name;
+        size_t size;
+    };
+
+    // Information about a read/write instruction, including the name of the function from which it was emitted.
+    struct ReadWriteLogEntry {
+        enum class AccessType { Read, Write };
+
+        uint32_t instructionAddress;
+        std::string functionName;
+        AccessType accessType;
+    };
+
+    // A WatchTreeNode contains the information necessary for displaying a type hierarchy in a tree.
+    struct WatchTreeNode {
+        std::string type;
+        std::string name;
+        size_t size = 0;
+        std::vector<WatchTreeNode> children;
+        std::vector<ReadWriteLogEntry> logEntries;
+    };
+
+    using StructFields = std::vector<FieldOrArgumentData>;
+    // Populates a node according to its type.
+    void populate(WatchTreeNode* node, std::unordered_map<std::string, StructFields>& structs_info);
+
+    // For an explanation of the meaning of addressOfPointer, see declaration of displayNode().
+    struct AddressNodeTuple {
+        uint32_t address;
+        bool addressOfPointer;
+        WatchTreeNode node;
+    };
+
+    struct DebuggerFunction {
+        std::string name;
+        std::vector<FieldOrArgumentData> arguments;
+    };
+
+    // Represents a value stored in a register, as opposed to one stored at an address in memory.
+    struct RegisterValue {
+        uint32_t value;
+        std::string type;
+        std::string name;
+        size_t size = 0;
+    };
+
+    /**
+     * Data types.
+     */
+
+    std::unordered_map<std::string, StructFields> m_structs;
+    std::vector<std::string> m_typeNames;
+
+    /**
+     * Watch.
+     */
+
+    std::vector<AddressNodeTuple> m_displayedWatchData;
+    std::vector<PCSX::Debug::Breakpoint*> m_watchBreakpoints;
+    std::unordered_map<uint32_t, std::array<uint8_t, 4>> m_disabledInstructions;
+    bool m_hex = false;
+    uint32_t m_newValue = 0;
+
+    /**
+     * Functions.
+     */
+
+    std::vector<uint32_t> m_functionAddresses;
+    std::unordered_map<uint32_t, DebuggerFunction> m_functions;
+
+    struct FunctionBreakpointData {
+        uint32_t id;
+        std::string functionName;
+        std::string callerName;
+        uint32_t callerAddress;
+        using ArgumentData = std::variant<AddressNodeTuple, RegisterValue>;
+        std::vector<ArgumentData> argData;
+    };
+    std::vector<FunctionBreakpointData> m_displayedFunctionData;
+    std::vector<PCSX::Debug::Breakpoint*> m_functionBreakpoints;
+    std::unordered_map<uint32_t, std::array<uint8_t, 8>> m_disabledFunctions;
+
+    // Returns the name of the function from which the instruction at the given address was emitted if found, an empty
+    // string otherwise.
+    std::string getFunctionNameFromInstructionAddress(uint32_t address);
```

Comment:
This probably should return an std::string_view instead of an std::string, to avoid allocating and copying strings.